### PR TITLE
Improve Product::getAttributesResume

### LINKED ISSUE #30533: Improve the Product::getAttributesResume
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Hello.
By working on a customer ticket, I discovered an issue into Product::getAttributesResume method.

Somewhere in our interface, a product_attribute should return an `attribute_designation` **color black**, but the method gave us **color green**.
After some reseachs I found there are more `$combinations` array entries than `$lang` array entries.
Fine, the first issue come from merchant data. We keep that in mind but it doesn't block the improvement.

**We could have be a great result even whith those wrong data**.

Here is why:
The foreach loop for `$lang` array fill `$combinations` array by using basic `$k`. 
So if the lenght of the two arrays dosn't match, there could be an error.

In my opinion, using auto_increment index is not the best way to match `$combinations` data with `$lang` data.

image

### Describe the solution you'd like

I think using 'id_product_attribute' as index of `$combinations` array solves the merchant issue, improves code and performances.

### Alternatives you've considered

_No response_

### Additional context

_No response_

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by StoreCommanderDev**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | PrestaShop < 1.7.8.0 <br>Improve code to correctly apply attribute lang informations to the right combinations<br>Using id_product_attribute as index instead of an auto increment array index for a right match.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #30533
| Related PRs       | https://github.com/PrestaShop/PrestaShop/pull/24089
| How to test?      | It's a technical improvement than a real error we can reproducing easily.<br>It's an add to this PR: https://github.com/PrestaShop/PrestaShop/pull/24089
| Possible impacts? | Admin Product Controller tab combinations

### CODE REVIEW

File: classes/Product.php [hunk 1]
```diff
@@ -2725,10 +2725,9 @@ public function getAttributesResume($id_lang, $attribute_value_separator = ' - '
             return false;
         }
 
-        $product_attributes = [];
-        foreach ($combinations as $combination) {
-            $product_attributes[] = (int) $combination['id_product_attribute'];
-        }
+        $combinations = array_column($combinations, null, 'id_product_attribute');
+
+        $product_attributes = array_keys($combinations);
```
Comment (zuk3975) [inline classes/Product.php]:
```suggestion
        $combinationIds = array_keys($combinations);
```
also don't forget to change bellow in the sql query.

**Review by zuk3975 — COMMENTED** [review]
just a little naming suggestion to make it more readable and its ok for me

**Review by nicosomb — APPROVED** [review]
OK for me

File: classes/Product.php (hunk 1 above)
Comment (StoreCommanderDev) [reply classes/Product.php]: Thanks you. It's done.

**Review by zuk3975 — APPROVED** [review]

**Comment by zuk3975** [comment]
@StoreCommanderDev could you describe how to test it?
At least to check that nothing breaks, like the combinations are still shown as they should (in whatever page that this might have effect)

**Comment by StoreCommanderDev** [comment]
> @StoreCommanderDev could you describe how to test it? At least to check that nothing breaks, like the combinations are still shown as they should (in whatever page that this might have effect)

In Combinations tab in Product Admin Controller, save correspondances between id_combination and attributes associated (ex: id_combination 11 is black , id_combination 12 is red and id_combination 13 is green)
After patch application check data are the same in same order

**Comment by zuk3975** [comment]
> > @StoreCommanderDev could you describe how to test it? At least to check that nothing breaks, like the combinations are still shown as they should (in whatever page that this might have effect)
> 
> In Combinations tab in Product Admin Controller, save correspondances between id_combination and attributes associated (ex: id_combination 11 is black , id_combination 12 is red and id_combination 13 is green) After patch application check data are the same in same order

Does it impact something in Front office? We don't care much about old product page in BO, because its gonna soon be replaced by new one (which is under feature flag now, but is close to finish).
So if you could provide some test scenario for Front office that would be much better.

**Comment by StoreCommanderDev** [comment]
I think there is no impact on Front Office, because this patch impact only the `attribute_designation` entry.

But the method `getAttributesResume` is in `ProductLazyArray` method `shouldShowOutOfStockLabel()` to know if a product need to show the out of stock label.

So, after patching code we can test if getting flag works
Example: if the quantity of combination 11 is 0, in front office product page if I switch between combinations when I click on the combination 11 it must be a message like `Product available with different options`

**Comment by zuk3975** [comment]
> I think there is no impact on Front Office, because this patch impact only the `attribute_designation` entry.
> 
> But the method `getAttributesResume` is in `ProductLazyArray` method `shouldShowOutOfStockLabel()` to know if a product need to show the out of stock label.
> 
> So, after patching code we can test if getting flag works Example: if the quantity of combination 11 is 0, in front office product page if I switch between combinations when I click on the combination 11 it must be a message like `Product available with different options`

@PrestaShop/committers  I think QA by dev should be enough, wdyt? Anyone willing to test it?

**Comment by StoreCommanderDev** [comment]
Does the "Waiting for author" tag can be removed?

**Review by sallemiines — COMMENTED** [review]
Hello @StoreCommanderDev 
Thanks for the PR ! It's LGTM QA :heavy_check_mark: 

Thank you ! :rocket:

**Comment by Hlavtox** [comment]
Thanks @StoreCommanderDev & @sallemiines! :-)

**Comment by StoreCommanderDev** [comment]
🥳  ouhou ! thank you all

**Pull request merged**